File: backend/src/main/java/com/rainier/portfolio/Ryg.java
/* (C) 2026 Rainier — internal use only. */
package com.rainier.portfolio;

/**
 * Deterministic project red/yellow/green health (v0.0.28). Mirrors the frontend {@code ryg.ts} rule
 * so the server-side portfolio and the client panels agree: GRAY when no open work; RED if anything is
 * blocked or the overdue ratio exceeds the threshold; YELLOW if some (but few) overdue; GREEN otherwise.
 */
public final class Ryg {

  /** A project goes RED above this overdue ratio (of its open tasks). */
  public static final double OVERDUE_RED_THRESHOLD = 0.3;

  private Ryg() {}

  public static String tier(int openCount, int overdueCount, int blockedCount) {
    if (openCount == 0) {
      return "GRAY";
    }
    if (blockedCount > 0) {
      return "RED";
    }
    double ratio = (double) overdueCount / openCount;
    if (ratio > OVERDUE_RED_THRESHOLD) {
      return "RED";
    }
    if (overdueCount > 0) {
      return "YELLOW";
    }
    return "GREEN";
  }

  /** Sort order red &lt; yellow &lt; green &lt; gray (worst first). */
  public static int order(String tier) {
    switch (tier) {
      case "RED":
        return 0;
      case "YELLOW":
        return 1;
      case "GREEN":
        return 2;
      default:
        return 3;
    }
  }
}

File: frontend/src/utils/ryg.ts
/**
 * v0.0.25 team-lead panel — pure red/yellow/green health rules + member-load thresholds. No React,
 * no I/O. Named-constant thresholds so the rules are auditable and unit-tested in isolation.
 */
import type { StatusTier } from './board';

export type RygTier = StatusTier; // red | yellow | green | gray

/** A project goes RED above this overdue ratio (of its open tasks). */
export const RYG_OVERDUE_RED_THRESHOLD = 0.3;
/** Member-load colour bands by open-task count. */
export const LOAD_YELLOW_MIN = 4;
export const LOAD_RED_MIN = 8;

/** Sort order red → yellow → green → gray. */
export const RYG_ORDER: Record<RygTier, number> = { red: 0, yellow: 1, green: 2, gray: 3 };

const CLOSED_TASK = new Set(['DONE', 'CANCELLED']);

/** An open task is anything not DONE/CANCELLED (TODO/IN_PROGRESS/BLOCKED). */
export function isOpenTaskStatus(status: string): boolean {
  return !CLOSED_TASK.has(status.toUpperCase());
}

export interface RygInput {
  openCount: number;
  overdueCount: number;
  anyBlocked: boolean;
}

/**
 * Project health: GRAY if no open work; RED if any open task is BLOCKED or the overdue ratio exceeds
 * the threshold; YELLOW if some (but few) overdue; GREEN if open with nothing overdue/blocked.
 */
export function ryg({ openCount, overdueCount, anyBlocked }: RygInput): RygTier {
  if (openCount === 0) return 'gray';
  if (anyBlocked) return 'red';
  const ratio = overdueCount / openCount;
  if (ratio > RYG_OVERDUE_RED_THRESHOLD) return 'red';
  if (ratio > 0) return 'yellow';
  return 'green';
}

/** Member load colour by open-task count: <=3 green, 4–7 yellow, >7 red. */
export function loadTier(openTasks: number): RygTier {
  if (openTasks >= LOAD_RED_MIN) return 'red';
  if (openTasks >= LOAD_YELLOW_MIN) return 'yellow';
  return 'green';
}
